DirectXTex/DirectXTexSwizzle.cpp

```diff
@@ -0,0 +1,285 @@
+        uint32_t res = 0;
+        for (uint32_t bb = 1; mask !=0; bb += bb)
+        {
+            if (val & mask & -mask)
+            {
+                res |= bb;
+            }
+            mask &= (mask - 1);
+        }
+        return res;
+    }
+#endif
+}
+
+
+_Use_decl_annotations_
+HRESULT DirectX::StandardSwizzle(const Image& srcImage, bool toSwizzle, ScratchImage& result) noexcept
+{
+    if (!IsValid(srcImage.format))
+        return E_INVALIDARG;
+
+    if (IsTypeless(srcImage.format) || IsPlanar(srcImage.format) || IsPalettized(srcImage.format))
+        return HRESULT_E_NOT_SUPPORTED;
+
+    HRESULT hr = result.Initialize2D(srcImage.format, srcImage.width, srcImage.height, 1, 1);
+    if (FAILED(hr))
+        return hr;
+
+    const uint8_t* sptr = srcImage.pixels;
+    if (!sptr)
+        return E_POINTER;
+
+    uint8_t* dptr = result.GetImages()[0].pixels;
+    if (!dptr)
+        return E_POINTER;
+
+    size_t bytesPerPixel = BitsPerPixel(srcImage.format) / 8;
+
+    uint32_t xBytesMask = 0b1010101010101010;
+    uint32_t yBytesMask = 0b0101010101010101;
+
+    size_t height = IsCompressed(srcImage.format) ? (srcImage.height + 3) / 4 : srcImage.height;
+    size_t width  = IsCompressed(srcImage.format) ? (srcImage.width  + 3) / 4 : srcImage.width;
+
+    if (toSwizzle)
+    {
+        size_t rowPitch = srcImage.rowPitch;
+        for (size_t y = 0; y < height; y++)
+        {
+            for (size_t x = 0; x < width; x++)
+            {
+                uint32_t swizzleIndex = deposit_bits(x, xBytesMask) + deposit_bits(y, yBytesMask);
+                size_t swizzleOffset = swizzleIndex * bytesPerPixel;
+
+                size_t rowMajorOffset = y * rowPitch + x * bytesPerPixel;
+
+                const uint8_t* sourcePixelPointer = sptr + rowMajorOffset;
+                uint8_t* destPixelPointer = dptr + swizzleOffset;
+                memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
+            }
+        }
+    }
+    else
+    {
+        size_t rowPitch = result.GetImages()[0].rowPitch;
+        for (size_t swizzleIndex = 0; swizzleIndex < (width * height); swizzleIndex++)
+        {
+            size_t swizzleOffset = swizzleIndex * bytesPerPixel;
+
+            uint32_t destX = extract_bits(swizzleIndex, xBytesMask);
+            uint32_t destY = extract_bits(swizzleIndex, yBytesMask);
+            size_t rowMajorOffset = destY * rowPitch + destX * bytesPerPixel;
+
+            const uint8_t* sourcePixelPointer = sptr + swizzleOffset;
+            uint8_t* destPixelPointer = dptr + rowMajorOffset;
+            memcpy(destPixelPointer, sourcePixelPointer, bytesPerPixel);
+        }
+    }
+
+    return S_OK;
+}
+
+_Use_decl_annotations_
+HRESULT DirectX::StandardSwizzle(const Image* srcImages, size_t nimages, const TexMetadata& metadata, bool toSwizzle, ScratchImage& result) noexcept
+{
+    HRESULT hr = result.Initialize2D(srcImages[0].format, srcImages[0].width, srcImages[0].height, nimages, 1);
```

Comment:
Ideally for the error state coming out of all DirectXTex functions, the ``ScratchImage`` should be empty. I usually do all parameter validation first, then create the result working space, and at any failure-point after that I explicitly call ``result.Release();``

Also, you can let `ScratchImage` do a lot of the work for you setting up complex surfaces.

```
if (!srcImages || !nimages || !IsValid(metadata.format))
        return E_INVALIDARG;

if (metadata.IsVolumemap() || IsTypeless(metadata.format) || IsPlanar(metadata.format) || IsPalettized(metadata.format))
        return HRESULT_E_NOT_SUPPORTED;
        
HRESULT hr = result.Initialize(metadata);
if (FAILED(hr)
    return hr;
```